### PR TITLE
[feat] Remove kube-rbac-proxy to expose metrics and replace it with current best practice

### PR DESCRIPTION
**Pull request by komer3**
Current way exposing metrics for the controller has changed and we no longer should be using kube-rbac-proxy. For more info: https://book.kubebuilder.io/reference/metrics.html

As part of this PR, I refactored the main() and moved flag parsing, setting up manager, setting up controllers, etc into their own functions for better readability and maintainability.

 

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -75,65 +96,76 @@ func init() {
...
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
+
+	setupLog.Info(fmt.Sprintf("CAPL version: %s", version.GetVersion()))
+
+	linodeClientConfig, dnsClientConfig := validateEnvironment()
+	mgr := setupManager(flags, linodeClientConfig, dnsClientConfig)
+
+	ctx := ctrl.SetupSignalHandler()
+	run := func(ctx context.Context) error {
+		o11yShutdown := setupObservabillity(ctx)
```
Comment (rahulait) [inline cmd/main.go:110]: Nitpick: fix spelling for `setupObservabillity` to `setupObservability`

Comment (komer3) [reply cmd/main.go:110]: oops good catch!

File: cmd/main.go [hunk 2]
```diff
@@ -143,142 +175,165 @@ func main() {
...
+	if mgr == nil {
+		setupLog.Error(errors.New("manager creation failed"), "manager is nil")
 		os.Exit(1)
 	}
 
-	if err = (&controller.LinodeClusterReconciler{
+	setupControllers(mgr, flags, linodeConfig, dnsConfig)
+
+	// Setup webhooks if enabled
+	if os.Getenv("ENABLE_WEBHOOKS") != "false" {
```
Comment (AshleyDumaine) [inline cmd/main.go:228]: probably want to use strconv.ParseBool here

**Comment by AshleyDumaine** [comment]
LGTM aside from the above minor comment

**Review by AshleyDumaine — DISMISSED** [review]

**Review by AshleyDumaine — DISMISSED** [review]

**Review by AshleyDumaine — DISMISSED** [review]

File: cmd/main.go (hunk 2 above)
Comment (komer3) [reply cmd/main.go]: Updated it!

File: config/network-policy/allow-metrics-traffic.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# This NetworkPolicy allows ingress traffic
+# with Pods running on namespaces labeled with 'metrics: enabled'. Only Pods on those
+# namespaces are able to gathering data from the metrics endpoint.
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  labels:
+    app.kubernetes.io/name: output-dir
```
Comment (rahulait) [inline config/network-policy/allow-metrics-traffic.yaml:8]: nit pick: I wonder if we want to add a label as `app.kubernetes.io/name: output-dir`

**Review by AshleyDumaine — DISMISSED** [review]

File: config/network-policy/allow-metrics-traffic.yaml (hunk 3 above)
Comment (komer3) [reply config/network-policy/allow-metrics-traffic.yaml]: updated!

**Review by rahulait — APPROVED** [review]
LGTM

**Pull request merged**